docs/flags.md

```diff
@@ -0,0 +1,130 @@
+---
+title: Flags for Tiltfile Config Management
+layout: docs
+---
+
+
+This doc describes how to use flags to maintain configuration of a developer instance. This is especially useful for teams that have both enough services that there are multiple ways you might want to run it, and a designated owner of the developer experience (often a "DevEx" team or principal engineer). We'll use "you" to mean that DevEx engineer and "your user" to mean a developer at your org that uses Tilt.
+
+## Goal
+Users can start Tilt in a specific mode with a single Tilt command. For instance, `tilt up consumer` will start just the services for consumer users (vs other modes, like `tilt up enterprise`). If they later run `tilt up` it can stay in the selected mode, and they can change the mode while tilt is running with `tilt flags enterprise`.
+
+This doc describes how to describe your setup to Tilt to enable these shorthands. We aim to make common cases trivial, while also supporting escape hatches that enable arbitrarily-complex cases.
+
+In order, we'll cover:
+*) Examples that cover common cases, showing both the Tiltfiles and accompanying usage.
+*) Describe how this is implemented. These details, especially around persistence, will let you understand how to write tooling that interacts with this system.
+*) Future work (we see how it fits in but don't plan to implement until there's concrete demand).
+
+## Examples
+
+### Run only some services
+Your app has many services, call them A, B, C, and D. You want users to be able to run only some of them.
+
+and you want your users to be able to run only a subset of them. E.g., `tilt up` will run all services, but running `tilt up a b d` will run only the resources a, b, and d, ignoring c. Each run of Tilt will start from scratch.
+#### Command-Lines
+* `tilt up`: run all services
+* `tilt up a b d`: run A, B, and D but not C
+* `tilt flags a b`: change running Tilt to run A and B but neither C nor D.
+* `tilt up` (a second time): run all services
+
+#### Tiltfile
+```python
+flags.define_string_list("to-run", args=True, sticky=False)
+cfg = flags.parse()
+
+flags.exclude_all_resources()
```

Comment:
Is there a reason the `resource` builtins are on `flags`, or is it just because we're adding them at the same time? What if I want to use them instead with `read_json` or `local`?

Reply:
Because it has to do with customization. I wasn't sure whether to call the module flags or config or args or custom. Open to better ideas.

Reply:
> yeah Matt, pretty sure in this proposal you could say cfg = read_json('config.json') -- is that what you're asking about?

right, what I'm saying is that it's weird to turn around and have to pass that to something in the `flags` namespace when I'm not using flags at all

> Because it has to do with customization. I wasn't sure whether to call the module flags or config or args or custom. Open to better ideas.

`config` seems fine, or even just putting it in the global namespace